### PR TITLE
Refactor WebServer and add unit tests

### PR DESCRIPTION
**Pull request by torbacz**
## Description

- Slightly refactored WebServer class in order to be able to create unit test
- Created basic unit tests for route matching

## Motivation and Context

- I want to introduce url parameters (like in full .net where you can put {id} in URL and it matches method parameters) and maybe serialization of body parameters into method objects.
- I want to create benchmarks for resolving routes but for that I need a lot of unit tests :) 

## How Has This Been Tested?

- unit tests

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [ ] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

File: nanoFramework.WebServer/WebServer.cs [hunk 1]
```diff
@@ -665,6 +641,43 @@ protected virtual void InvokeRoute(CallbackRoutes route, HttpListenerContext con
             route.Callback.Invoke(null, new object[] { new WebServerEventArgs(context) });
         }
 
+        /// <summary>
+        /// Logs a message.
+        /// </summary>
+        /// <param name="message">The message to be logged.</param>
+        protected virtual void Log(string message)
```
Comment (Ellerbach) [inline nanoFramework.WebServer/WebServer.cs:648]: not sure it's optimal because at compilation Debug is removed when you're building in release. But the function will stay and the string transformation will stay as well. I'm not super sure the compiler  will remove everything. And as C# does not support "inline" function  in the sense of C++, I don't think this will really optimise the code.
See the MQTT lib, to avoid this issue because the log are formatted and requires a bit of love, I4ve added some #if DEBUG in the code whenever there is something to debug and format. So it's just totally skipped when not going thru.
Note for later: We should definitely test this behavior with the benchmark lib. Some function of this kind, empty inside, with string parameters and transformed.

File: nanoFramework.WebServer/WebServer.cs [hunk 2]
```diff
@@ -665,6 +641,43 @@ protected virtual void InvokeRoute(CallbackRoutes route, HttpListenerContext con
...
+                return;
+            }
+
+            try
+            {
+                context.Response.Close();
+            }
+            catch
+            {
+                // Nothing on purpose
```
Comment (Ellerbach) [inline nanoFramework.WebServer/WebServer.cs:668]: Note for another PR: we should add an event or equivalent here in this function all up. Because it may be due to a wifi disconnection or equivalent and the device will continue to run but the network won't and the webserver will stop.

File: tests/nanoFramework.WebServer.Tests/WebServerTests.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        [DataRow("GET", "/API/TEST", "/api/test")]
+        [DataRow("POST", "/API/TEST", "/api/test")]
+        [DataRow("PUT", "/API/TEST", "/api/test")]
+        [DataRow("PATCH", "/API/TEST", "/api/test")]
+        [DataRow("DELETE", "/API/TEST", "/api/test")]
+        [DataRow("GET", "/api/test", "/API/TEST")]
+        [DataRow("POST", "/api/test", "/api/test")]
+        [DataRow("PUT", "/api/test", "/API/TEST")]
+        [DataRow("PATCH", "/api/test", "/API/TEST")]
+        [DataRow("DELETE", "/api/test", "/API/TEST")]
```
Comment (Ellerbach) [inline tests/nanoFramework.WebServer.Tests/WebServerTests.cs:61]: Can you add couple of longer URL with parameters just to see how it does react?

File: tests/nanoFramework.WebServer.Tests/WebServerTests.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
+using nanoFramework.TestFramework;
```
Comment (Ellerbach) [inline tests/nanoFramework.WebServer.Tests/WebServerTests.cs:1]: you're missing the headers ;-)

File: tests/nanoFramework.WebServer.Tests/WebServerTests.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        // IsRouteMatch CaseSensitive true/false
+
+        [TestMethod]
+        public void IsRouteMatch_Should_ReturnFalseForEmptyMethod()
+        {
+            // Arrange
+            var route = new CallbackRoutes()
+            {
+                Method = "GET",
+                Route = "/api/test"
```
Comment (Ellerbach) [inline tests/nanoFramework.WebServer.Tests/WebServerTests.cs:19]: same here, worth having a test with query parameters as well

File: tests/nanoFramework.WebServer.Tests/WebServerTests.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        
+        [TestMethod]
+        public void IsRouteMatch_Should_ReturnFalseForNotMatchingMethod()
+        {
+            // Arrange
+            var route = new CallbackRoutes()
+            {
+                Method = "GET",
+                Route = "/api/test"
```
Comment (Ellerbach) [inline tests/nanoFramework.WebServer.Tests/WebServerTests.cs:36]: and here as well

**Review by Ellerbach — COMMENTED** [review]
Looks good all up and right direction. Thanks for putting love in the web server. There are few things I4ve seen using it recently in an advance project.  So some notes for a further PR, not for this one.

File: nanoFramework.WebServer/WebServer.cs (hunk 1 above)
Comment (torbacz) [reply nanoFramework.WebServer/WebServer.cs:648]: The idea was to be able to replace "Debug.Writeline" with own logger. For example, i'm using "serilog" like logger which is able to write logs into multiple sinks (like Console or SDcard).
But maybe better idea is to introduce events? The we can listen to those events?

File: tests/nanoFramework.WebServer.Tests/WebServerTests.cs (hunk 3 above)
Comment (torbacz) [reply tests/nanoFramework.WebServer.Tests/WebServerTests.cs:61]: Sure, I'll add few more. If you have any edge cases to check, leave me a note and I'll add more tests.

File: nanoFramework.WebServer/WebServer.cs (hunk 2 above)
Comment (torbacz) [reply nanoFramework.WebServer/WebServer.cs:668]: Few more from me:
Maybe we should introduce, few more default behaviours for WebServer:
- When there is no route, we could return 404
- When there is error in method invoked - return 500

File: tests/nanoFramework.WebServer.Tests/WebServerTests.cs (hunk 3 above)
Comment (Ellerbach) [reply tests/nanoFramework.WebServer.Tests/WebServerTests.cs:61]: I think testing withn things like this can be great:
* /api/test?id=1234
* /api/test/?id=1234
* /api/test?
* /api/api/test
* /api/api/test?
* /api/api/test/?id=1234
* /api/api/test?id=1234
* /tagada/ap/test

things like this!

File: nanoFramework.WebServer/WebServer.cs (hunk 1 above)
Comment (Ellerbach) [reply nanoFramework.WebServer/WebServer.cs:648]: I like the logger and we have a standardize logger with API aligned with .NET. So we could use this one. Key element here as well is performance. I'm running some tests to see how the log function behave

Comment (Ellerbach) [reply nanoFramework.WebServer/WebServer.cs:648]: OK, just ran the tests:

```csharp
public void DoWithLog()
{
    int a = 0;
    for (int i = 0; i < 1_000; i++)
    {
        a++;
        Log($"Some text and a is now: {a}");
    }
}

public void DoWithDebug()
{
    int a = 0;
    for (int i = 0; i < 1_000; i++)
    {
        a++;
        Debug.WriteLine($"Some text and a is now: {a}");
    }
}

public void DoWitNothing()
{
    int a = 0;
    for (int i = 0; i < 1_000; i++)
    {
        a++;
    }
}

public void Log(string s)
{
    Debug.WriteLine(s);
}
```

Results I ran each of those 10 times), all in Release:

Do with Log: 00:01:28.2310360
Do with Debug : 00:00:00.0457140
Do with Nothing : 00:00:00.0455600

So as expected, there is a penalty for using a log function even if it's empty! Building in release is the equivalent of removing the Debug.WriteLine. As you can see in the test, even the empty function is called! So you have a penalty creating the string.

Bottom line:
* using the #if DEBUG is the way to go
* inside the block, you can use anything you want
* if you really want to use a logger, then the PN532 way is the way to go:
  *  check before calling the format function is the log level is higer than what you need to log.
 ``csharp
if (loglevel >= refLogLevel)
{
  Log($"Whatever you want to format {anything_here}");
]
 ```
This is going to improve performances!

File: nanoFramework.WebServer/WebServer.cs (hunk 2 above)
Comment (Ellerbach) [reply nanoFramework.WebServer/WebServer.cs:668]: Great list! Let's go for this PR first :-) And let's see how to continue to improve it ;-)

File: nanoFramework.WebServer/WebServer.cs (hunk 1 above)
Comment (torbacz) [reply nanoFramework.WebServer/WebServer.cs:648]: OK, for now rolling back to "Debug.Writeline"

File: tests/nanoFramework.WebServer.Tests/WebServerTests.cs (hunk 3 above)
Comment (torbacz) [reply tests/nanoFramework.WebServer.Tests/WebServerTests.cs:61]: Is the "/api/api/test/?id=1234" valid URL? (/ and then ?)
I'm asking because current implementation fails to resolve it.

File: tests/nanoFramework.WebServer.Tests/WebServerTests.cs (hunk 4 above)
Comment (torbacz) [reply tests/nanoFramework.WebServer.Tests/WebServerTests.cs:6]: done

File: nanoFramework.WebServer/WebServer.cs (hunk 2 above)
Comment (torbacz) [reply nanoFramework.WebServer/WebServer.cs:668]: And one more:
- If we are using controllers do we need "CommandReceived" event? Maybe we should decouple logic and create two separate web servers - one with controllers and one with "CommandReceived". Or maybe just get rid of CommandReceived? I'm not sure if full .net supports such case as "CommandReceived"?

File: tests/nanoFramework.WebServer.Tests/WebServerTests.cs (hunk 3 above)
Comment (torbacz) [reply tests/nanoFramework.WebServer.Tests/WebServerTests.cs:63]: Added few more tests for routes.
Later I'll add few more.

File: tests/nanoFramework.WebServer.Tests/WebServerTests.cs (hunk 5 above)
Comment (torbacz) [reply tests/nanoFramework.WebServer.Tests/WebServerTests.cs:21]: Here i'm testing if the method will return early for empty method name. I don't thing query params here are required.

File: tests/nanoFramework.WebServer.Tests/WebServerTests.cs (hunk 6 above)
Comment (torbacz) [reply tests/nanoFramework.WebServer.Tests/WebServerTests.cs:38]: Here i'm testing if the method will return early for empty method name. I don't thing query params here are required.

File: tests/nanoFramework.WebServer.Tests/WebServerTests.cs (hunk 3 above)
Comment (torbacz) [reply tests/nanoFramework.WebServer.Tests/WebServerTests.cs:63]: I think that's all for now. I'll extend tests when working on improving web server (better tests would be - create controller , so the "callbackroute" is created based on that and the check if route matches)

Comment (Ellerbach) [reply tests/nanoFramework.WebServer.Tests/WebServerTests.cs:63]: > Is the "/api/api/test/?id=1234" valid URL? (/ and then ?)
> I'm asking because current implementation fails to resolve it.

Well, never tested that one and I asked Bing chat for some help.

"Yes, that is a valid URL. The characters / and ? can be together like that, as they have different meanings in different parts of the URL.

The / character is used to separate path segments in the URL. In your example, the path is /api/api/test/, which means that the resource you are requesting is located under the test directory, which is under the api directory, which is under the api directory, which is under the root directory of the server.

The ? character is used to separate the query string from the rest of the URL. The query string is a set of key-value pairs that provide additional information to the server. In your example, the query string is id=1234, which means that you are passing the value 1234 for the key id to the server.

The query string is optional and can be omitted if you don’t need to pass any information to the server. However, the / character is mandatory and cannot be omitted, as it indicates the end of the path and the start of the query string.

If you want to learn more about the syntax and semantics of URLs, you can check out these resources:

[Web Naming and Addressing Overview (URIs, URLs, …)](https://www.w3.org/Addressing/)
[Uniform Resource Locators - World Wide Web Consortium (W3C)](https://www.w3.org/Addressing/URL/url-spec.html)
[RFC 1738 - Uniform Resource Locators (URL) - IETF Datatracker](https://datatracker.ietf.org/doc/html/rfc1738)
I hope this helps. If you have any other questions, please let me know. 😊"

So, we may have to treat it as a valid URL. For the time being, we can ignore it and add it to the list of improvements. It's definitely a edge case!

File: nanoFramework.WebServer/WebServer.cs (hunk 2 above)
Comment (Ellerbach) [reply nanoFramework.WebServer/WebServer.cs:659]: We need that one with the controllers. I've been using this one to trap favicon, style sheets and other elements. So this is a necessary one, regardless of using controllers or not.

File: tests/nanoFramework.WebServer.Tests/WebServerTests.cs (hunk 5 above)
Comment (Ellerbach) [reply tests/nanoFramework.WebServer.Tests/WebServerTests.cs:21]: ok, fair enough

**Review by Ellerbach — APPROVED** [review]
Looks good! Thanks.

**Comment by Ellerbach** [comment]
@torbacz resolve the comments if you don't have anything else to add. We can then merge this one and move to other improvments.

File: nanoFramework.WebServer/WebServer.cs (hunk 2 above)
Comment (torbacz) [reply nanoFramework.WebServer/WebServer.cs:659]: Ahh, Ok.

**Comment by Ellerbach** [comment]
> And I've create a TODO list for future

Shall we create a specific project for the web server? I think I can set this up in nanoFramework org. So we can track what needs to be done with some description in cards. And the one working on something just move the card. So we don't double the efforts. Thoughts?

**Comment by torbacz** [comment]
@Ellerbach Sure, please create one and send me a link. I'll move list there.

**Comment by josesimoes** [comment]
@Ellerbach yes please. We should be using projects a lot more than what we are today. 
It's great for keeping track of ideas and tasks without cluttering repos with issues, which aren't really issues.

**Comment by Ellerbach** [comment]
@torbacz and @josesimoes the project is now created, I've invited @torbacz to the project. We (José and I) can invite anyone. So let's promote it whenever there is work to do and to take. I'll add the improvements I've seen.
